docs/design.md

```diff
@@ -0,0 +1,128 @@
+# Overview
+
+Concord-bft is an implementation of the [SBFT
+Protocol](https://arxiv.org/abs/1804.01626). SBFT implements State Machine
+Replication (SMR) in the byzantine model. It is a generic protocol, with
+pluggable interfaces for communication, storage, and state transfer. This
+document covers the implementation of concord-bft. When the implementation
+differs from the paper, it will be mentioned explicitly.
+
+It is important to note that this document covers the code as it exists **NOW**.
+When the code changes substantially this document should be updated.
+
+# Public Interfaces
+
+Concord-bft is a library meant to be used by application code. The external user
+API for the core engine is described by a [set of structs and abstract
+interfaces](../bftengine/include/bftengine). The file names are pretty self
+explanatory. Important structures will be mentioned explicitly below.
+
+All user interfaces live in the `bftEngine` namespace.
+
+# Clients
+A default client,
+[SimpleClient](../bftengine/include/bftengine/SimpleClient.hpp), is included
+with concord-bft. It is just one manner of implementing a client, and is
+implemented in C++ utilizing the `ICommunication` and `IReceiver` interfaces
+below. Additional clients can be implemented by utilizing the ClientMessages
+described in the next subsection. In fact, we have a [client in
+Python](../util/pyclient/bft_client.py) that we use for testing that does just
+that.
+
+## Client Messages
+Client request and reply messages are prefixed with a `ClientRequestHeader` or
+`ClientReplyHeader` respectively. Each header, like all other messages in
+concord-bft is described as a packed C struct, meaning that there are no padding
+bytes between struct fields. This is accomplished via enclosing the structure
+definitions between `#pragma pack(push,1)` and `#pragma pack(pop)` lines.
+Packing ensures that the size of the structure is the sum of the size of the
+fields, and allows them to be sent over the wire and interpreted
+deterministically, given both machines have the same endianess.
+
+The primary benefit of this approach is that it is optimal in time and space
+requirements. Structs are exactly as large as they need to be, and there is no
+need to serialize and deserialize them, as they are operated on in memory
+identically to how they are sent on the wire. It is also very easy to
+implement and understand.
+
+The drawbacks are:
+ * Endianness is not taken into consideration, so a struct sent from a big
+   endian machine will be incorrectly interpreted on a little endian machine,
+   and vice versa.
+ * Messages field access may be slow, as it may not be on an aligned boundary.
+ * Messages are fixed. New fields cannot be added to messages and remain
+   backwards compatible.
+
```

Comment:
Perhaps this document should just describe the system without providing too much opinion. We can move the advantages and drawbacks to a different document that discusses design choices and roadmap.

Reply:
Good point!